### PR TITLE
Add CI: Format, Lint & Weak Type Check

### PR DESCRIPTION
**Pull request by bh2smith**
This PR adds a workflow file that runs black pylint & mypy.

The reason for the invasive change on the primary function was the deep nesting compliant from pylint. Before there were three nested loops (on shards, receipt_execution_outcomes, logs) which has been broken into three `process_XXX` methods.

Technically it would be possible to introduce this change without the big refactor (by simply adding a pylint disable to this particular message

```
R1702: Too many nested blocks (7/5) (too-many-nested-blocks)
``` 

and following up with another PR that addresses this only.

If accepted, I would proceed with 
1. adding a more strict type check (i.e. mypy --strict) and fixing the 
<details>
<summary>related issues</summary>

```sh
gas_station_event_indexer.py:23: error: Function is missing a return type annotation  [no-untyped-def]
gas_station_event_indexer.py:73: error: Missing type parameters for generic type "dict"  [type-arg]
gas_station_event_indexer.py:86: error: Missing type parameters for generic type "dict"  [type-arg]
gas_station_event_indexer.py:104: error: Returning Any from function declared to return "dict[Any, Any] | None"  [no-any-return]
gas_station_event_indexer.py:107: error: Function is missing a return type annotation  [no-untyped-def]
gas_station_event_indexer.py:112: error: Function is missing a return type annotation  [no-untyped-def]
gas_station_event_indexer.py:121: error: Function is missing a return type annotation  [no-untyped-def]
gas_station_event_indexer.py:131: error: Missing type parameters for generic type "dict"  [type-arg]
gas_station_event_indexer.py:153: error: Function is missing a return type annotation  [no-untyped-def]
gas_station_event_indexer.py:153: error: Missing type parameters for generic type "dict"  [type-arg]
gas_station_event_indexer.py:169: error: Function is missing a return type annotation  [no-untyped-def]
```
</details>
2. Adding unit tests to the new smaller methods.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,3 +2,8 @@ asyncio==3.4.3
 near-lake-framework==0.0.7
 requests==2.31.0
 toml==0.10.2
+
+black
```
Comment (anthony-near) [inline requirements.txt]: please include the versions for the new requirements.

**Review by anthony-near — CHANGES_REQUESTED** [review]
overall looks good. Just add versioning for packages added to requirements.txt

File: requirements.txt (hunk 1 above)
Comment (bh2smith) [reply requirements.txt]: Done.

File: requirements.txt [hunk 2]
```diff
@@ -2,3 +2,9 @@ asyncio==3.4.3
 near-lake-framework==0.0.7
 requests==2.31.0
 toml==0.10.2
+
+black>=24.4.0
+pylint>=3.1.0
+mypy>=1.9.0
+types-toml>=0.10.8.20240310
+types-requests>=2.31.0.20240406
```
Comment (bh2smith) [inline requirements.txt]:
```suggestion
types-requests>=2.31.0.20240406

```

**Review by anthony-near — APPROVED** [review]

**Pull request merged**